### DISCUSSION TITLE
Run coc.nvim on Bun, and failed

### DISCUSSION THREAD

**Discussion by jackielii** [neoclide/coc.nvim#4767] [Show and tell]
Hi,

Since bun has caught quite a bit of attention, I figured why not try running coc.nvim using Bun? And here is the journey and spoilers: I failed

I started with `bun esbuild.js`. It obviously worked because it just uses esbuild. However, since bun's [build API](https://bun.sh/docs/bundler#api) is borrowed from esbuild, so making a `bunbuild.js` is simple enough and it also worked.

So that was a good start. Next let's run the bundled js directly with `bun build/index.js`, and there is our first error:
![image](https://github.com/neoclide/coc.nvim/assets/360983/00b20df9-6f2c-4b23-b018-c92faeb1248a)

The code is inside [factory.ts](https://github.com/neoclide/coc.nvim/blob/06d5dfdee5bf9a90bf181d2e28723ccef685516d/src/util/factory.ts#L114) and reading through it, this is the extension loading mechanism. It uses Node's internal `Module._compile` API, in the newest version, it lives under [lib/internal/modules/cjs/loader.js](https://github.com/nodejs/node/blob/f16f41c5b3cf2345abc7bd4d5e951a43479ce4a7/lib/internal/modules/cjs/loader.js#L1383). Although depending on an internal API is not very good, but it seems [others does the same: jsonpath](https://github.com/dchester/jsonpath/blob/c1dd8ec74034fb0375233abb5fdbec51ac317b4b/lib/aesprim.js#L15).

OK, I was about to give up: Bun says it's compatible with Node, with lots of asterisks. Understandably this is not public API since it's prefixed with `_`. But just as I was moving away from it and refresh the page. The issue got closed by a PR that adds the `_compile` method: https://github.com/oven-sh/bun/pull/5840! Feeling rather lucky, I ran `bun upgrade --canary`, and continued...

So reading the `factory.ts`, it seems that the `getProtoWithCompile` is only used to extract the `_compile` method from `Module` so that it can be monkey patched in the `createSandbox` function. I don't fully get why this approach was taken. But it seems to rely on how `require` and `_compile` works internally in Nodejs's Module implementation. I don't know if this will work in Bun as Bun uses a native implementation in Zig, and it's not attached to the `prototype` Although this shouldn't stop us as we're not loading any extension yet. Let's just get this part happy, we won't hit it yet. To work around it, I just modify the top level  `ModuleProto` with:

```js
const ModuleProto = new Module('fake')
```

Re-run `bun build/index.js` now gives some text:
![image](https://github.com/neoclide/coc.nvim/assets/360983/8d28a4a8-89dd-4953-a4d4-0a056375c01f)

Since coc.nvim uses stdin and stdout to communicate with neovim, this is good news, it looks like the msgPack messages. I eagerly fired up nvim with minimal config:

```vim
set nocompatible
let g:coc_node_path = 'bun'
" let g:node_client_debug = 1
let g:coc_node_args = []
let $NODE_CLIENT_LOG_FILE = '/tmp/coc.log'
set runtimepath^=$PWD
filetype plugin indent on
syntax on
set hidden
```

Importantly `let g:coc_node_args = []` is needed because bun doesn't have ` --no-warnings` which is the default option added to the start up args. Then error:

![image](https://github.com/neoclide/coc.nvim/assets/360983/00d49133-ae69-42eb-87c7-0c5f8ebbd2ca)

At this point, I don't have any clue what is going on. So I figured I'll park this for now and share it just in case someone else find this interesting and want to continue

**Comment by chemzqm** [comment]
It's expected.  An internal method of the `module` module is hacked to load extensions https://github.com/neoclide/coc.nvim/blob/06d5dfdee5bf9a90bf181d2e28723ccef685516d/src/util/factory.ts#L151

**Comment by jackielii** [comment]
Some further findings forgot to share:

The full error message shown after running above is:

![image](https://github.com/neoclide/coc.nvim/assets/360983/06a814db-2c90-4812-8dd0-d4e6469f2a43)

It looks like the problem lies within the neovim's node-client. So I used this program to check (borrowed from [neovim node-client](https://github.com/neovim/node-client#attach)):

```javascript
const cp = require('child_process')
const attach = require('neovim').attach // replaced neovim with @chemzqm/neovim to cross check

const nvim_proc = cp.spawn('nvim', ['-u', 'NONE', '-N', '--embed'], {})

// Attach to neovim process
;(async function () {
  const nvim = await attach({ proc: nvim_proc })
  nvim.command('vsp')
  nvim.command('vsp')
  nvim.command('vsp')
  const windows = await nvim.windows

  // expect(windows.length).toEqual(4);
  // expect(windows[0] instanceof nvim.Window).toEqual(true);
  // expect(windows[1] instanceof nvim.Window).toEqual(true);

  nvim.window = windows[2]
  const win = await nvim.window

  // expect(win).not.toEqual(windows[0]);
  // expect(win).toEqual(windows[2]);

  const buf = await nvim.buffer
  // expect(buf instanceof nvim.Buffer).toEqual(true);

  const lines = await buf.lines
  // expect(lines).toEqual(['']);

  await buf.replace(['line1', 'line2'], 0)
  const newLines = await buf.lines
  // expect(newLines).toEqual(['line1', 'line2']);

  nvim.quit()
  // nvim_proc.disconnect()
})()
```

So as shown above, I ran the script with `neovim` package and `@chemzqm/neovim` package, and compared the result. The `neovim` package exited without error, the `@chemzqm/neovim` package showed error:

![image](https://github.com/neoclide/coc.nvim/assets/360983/d359da5c-9a76-40c6-ba80-d4b286aebe28)

Lastly I replaced all reference of `@chemzqm/neovim` to `neovim`, rebuilt, and re-run `nvim -u mini.vim` I get:
![image](https://github.com/neoclide/coc.nvim/assets/360983/a82a5b93-b623-4504-9677-01cca7d7ea66)

Albeit the error, `checkhealth` reports it's running:
![image](https://github.com/neoclide/coc.nvim/assets/360983/de9c2e6f-85d3-45d7-b651-3cfb1428673d)

Although I don't think it's actually working as none of the commands works. e.g. `CocInfo`

**Comment by meinzer1899** [comment]
Hey, any updates on this? `bun` 1.1.2 got [released](https://bun.sh/blog/bun-v1.1.2) recently.

**Comment by chemzqm** [comment]
It's likely not hard to run coc.nvim itself wtih bun, but we have some hacks with module to load extensions https://github.com/neoclide/coc.nvim/blob/master/src/util/factory.ts, don't know how to achieve that with bun.

**Comment by briancarbone** [comment]
I set ModuleProto to a fake module since I don't use extensions (I just configure manually) and it actually works perfectly fine with the bun runtime; I am not running into an fstat error with Vim 9.1.1355. I can't say I notice much of a difference compared to Node 23, but perhaps under certain conditions (very large projects or files), it could matter.

It reports itself as Node 22.6, but I can confirm bun is running and not node.

```
brian      64547  0.6  0.1 74162652 95900 ?      Ssl  11:21   0:00 /nix/store/86cjc3drywmhb3mhzy9kp4nknmrp8pvi-bun-1.2.10/bin/bun --no-warnings /nix/store/j1mfdlvg4px2bmf4cqn37g42v5bjwnma-vimplugin-coc.nvim/build/index.js
```

```
## versions

vim version: VIM - Vi IMproved 9.1 9011355
node version: v22.6.0
coc.nvim version: 0.0.82-master
coc.nvim directory: /nix/store/j1mfdlvg4px2bmf4cqn37g42v5bjwnma-vimplugin-coc.nvim
term: tmux
platform: linux

## Log of coc.nvim

2025-04-30T11:21:24.225 INFO (pid:64547) [plugin] - coc.nvim initialized with node: v22.6.0 after 93
2025-04-30T11:21:24.231 INFO (pid:64547) [services] - LanguageClient nixd state change: stopped => starting
2025-04-30T11:21:24.234 INFO (pid:64547) [language-client-index] - Language server "languageserver.nixd" started with 64598
2025-04-30T11:21:24.263 INFO (pid:64547) [services] - LanguageClient nixd state change: starting => running
2025-04-30T11:21:24.266 INFO (pid:64547) [services] - service languageserver.nixd started
2025-04-30T11:25:27.311 INFO (pid:64547) [attach] - receive notification: showInfo []
```
